### PR TITLE
dev/core#4567 Add `status`, `start_date` & `end_date` columns to `civicrm_mailing` table

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Add additional fields to civicrm_mailing table to allow job deletion

https://lab.civicrm.org/dev/core/-/issues/4567

Before
----------------------------------------
In order to find out the status of a mailing it is necessary to figure out the status of all the mailing jobs attached to it. On a large database this can be a big query. It also means we are looking at what should be temporary records (mailing job is about the process of sending) to get permanently interesting values

After
----------------------------------------
The start date, end date & status are added to the mailing record

Technical Details
----------------------------------------
I was on the fence about `status` (same as mailing_job) and `status_id` - the former seemed simpler so it is in place atm

Comments
----------------------------------------

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I tested this on our staging & it runs quickly - the mailing queue one is a bit slower but not insane

**Comment by mlutfy** [comment]
+1 from me. We do monitoring for stuck mailing, so it means checking against the job table at regular intervals (https://lab.civicrm.org/extensions/symbiotic/-/blob/master/CRM/Symbiotic/Utils/Check/MailingsPending.php)

This only adds the field, it does not update it, right?

**Comment by eileenmcnaughton** [comment]
@mlutfy that's correct - I was going to tackle the populating next - it's pretty confusing!

**Comment by mlutfy** [comment]
There maybe be a few people confused by the presence of the fields, but not being used, but I feel like delaying this is not worth losing the momentum to move forward.

(added merge-ready, in case anyone else has another opinion)

**Comment by colemanw** [comment]
@eileenmcnaughton if it's not too much trouble you could keep building the feature on this same branch and we can merge the PR when you get the fields populated.

**Comment by eileenmcnaughton** [comment]
To be honest I'm feeling a bit broken on understanding the code on this one - I might get there but I kinda want to get the one that populates the other fields merged (https://github.com/civicrm/civicrm-core/pull/27557)  before I tackle this because that part has the fields & the backfill already.

OTOH if I can figure this part out that's when we can simplify all those queries.  I *think* that flexmailer & core have duplicate code going on but I'm struggling to follow both sides of the code

**Comment by totten** [comment]
> (@eileenmcnaughton) I think that flexmailer & core have duplicate code going on but I'm struggling to follow both sides of the code

I believe that's correct - because CiviSMS still runs through the old one.

> To be honest I'm feeling a bit broken on understanding the code on this one - I might get there but I kinda want to get the one that populates the other fields merged (https://github.com/civicrm/civicrm-core/pull/27557) before I tackle this because that part has the fields & the backfill already.

OK, so that other PR was merged.

I agree that one complication is the existence of two delivery subsystems that touch this data. (*Flexmailer ~= SMTP; BAO-Engine ~= SMS*), but I think there's another complication which is more important - you need to condense/reduce information about many different status-records (from `civicrm_mailing_job`) into one status-record (in `civicrm_mailing`).  The logic which manages the jobs deals with a lot of error-scenarios and configuration-flags, so... well... it's not [easy like the smooth lilt of Lionel Richie's music](https://www.youtube.com/watch?v=gq-sDarXys8).

Suggestion: First, write the function to summarize the `civicrm_mailing_job`s and update the `civicrm_mailing`. I don't know if there's something to borrow already, but here's a brainstorm to demonstrate this kind of logic:

https://gist.github.com/totten/1542d0d0769d89b07cc101497fcdeaf3#file-updatemailingstatus-php-L41-L59

Once you have a summarization function, then it's just a matter of hooking into both mailers to call it.

This approach means that you don't need to understand the minutiae of building/splitting/updating/executing/resuming/finishing each specific `civicrm_mailing_job`. It should be easier to focus on using the existing data...

**Comment by eileenmcnaughton** [comment]
@totten thanks for that - I'll try to work through it - but #27557 actually isn't merged yet - I think you thought it was

**Comment by totten** [comment]
@eileenmcnaughton Ooh, I misread. You're right.

**Comment by eileenmcnaughton** [comment]
This is stale now & missed 5.67 so I'm gonna close it out for now & come back to it if I can face it

File: xml/schema/Mailing/Mailing.xml [hunk 1]
```diff
@@ -394,6 +394,47 @@
...
+    </html>
+    <readonly>true</readonly>
+  </field>
+  <field>
+    <name>status</name>
+    <title>Mailing Status</title>
+    <type>varchar</type>
+    <length>12</length>
+    <comment>The status of this mailing</comment>
+    <default>'Scheduled'</default>
```
Comment (colemanw) [inline xml/schema/Mailing/Mailing.xml:429]: Shouldn't this be
```suggestion
    <default>'Draft'</default>
```

File: CRM/Mailing/DAO/Mailing.php [hunk 2]
```diff
@@ -1195,6 +1222,84 @@ public static function &fields() {
...
+          'default' => 'Scheduled',
+          'table_name' => 'civicrm_mailing',
+          'entity' => 'Mailing',
+          'bao' => 'CRM_Mailing_BAO_Mailing',
+          'localizable' => 0,
+          'html' => [
+            'type' => 'Select',
+          ],
+          'pseudoconstant' => [
+            'callback' => 'CRM_Core_SelectValues::getMailingJobStatus',
```
Comment (colemanw) [inline CRM/Mailing/DAO/Mailing.php:1298]: I think we'll need a new pseudoconstant function to include the Draft status.

File: xml/schema/Mailing/Mailing.xml (hunk 1 above)
Comment (eileenmcnaughton) [reply xml/schema/Mailing/Mailing.xml:429]: hmm - yeah maybe - seems like a sensile default

File: CRM/Mailing/DAO/Mailing.php (hunk 2 above)
Comment (colemanw) [reply CRM/Mailing/DAO/Mailing.php:1298]: Or just add "Draft" to the existing function? Probably wouldn't hurt anything 🤷🏻

**Comment by eileenmcnaughton** [comment]
@colemanw this is stale again - I'm not sure if it's worth rebasing it - the only way I'm ever gonna get there on this one is to chip away piece by piece at it - but every time I look at it the next piece is dealing with the fact the db -add has gone stale.

I suspect that the setting the 'start_date' might not be that hard to do in a hook on MailingJob - well potentially they could all be done in that hook

**Comment by eileenmcnaughton** [comment]
The comments by @mlutfy about losing momentum were definitely prophetic here- I figured out the next steps ages ago but didn't get round to updating this to the latest civi-version so I could start on those next steps - in part cos I was waiting for the DAO complete. Anyway I've updated it now so hopefully we can merge & I can do the last bit & wrap this up

**Comment by mattwire** [comment]
@eileenmcnaughton Test fails relate I think (it's trying to validate a NULL value)

**Comment by eileenmcnaughton** [comment]
woohoo it passed - I've put the upgrade script into a separate PR as that will take some wrangling & would be better to get this merged

**Comment by eileenmcnaughton** [comment]
@aydun @colemanw just flagging that with this merged the Mailing screen would be a good candidate for AdminUI SK replacement as the query is now simple

**Comment by pradpnayak** [comment]
@mattwire is there any upgrade script to preset the values for start and end date for historical records?

**Comment by eileenmcnaughton** [comment]
@pradpnayak @mattwire this is the upgrade script - https://github.com/civicrm/civicrm-core/pull/30458 - I'm doing some more testing on it against the rc & would be super keen to get any other testing input

**Pull request merged**